source/common/upstream/health_checker_impl.cc

```diff
@@ -586,5 +595,297 @@ RedisHealthCheckerImpl::HealthCheckRequest::HealthCheckRequest() {
+  if (!client_) {
+    Upstream::Host::CreateConnectionData conn = host_->createConnection(parent_.dispatcher_);
+    client_ = parent_.createCodecClient(conn);
+    client_->addConnectionCallbacks(connection_callback_impl_);
+  }
+
+  request_encoder_ = &client_->newStream(*this);
+  request_encoder_->getStream().addCallbacks(*this);
+
+  auto headers_message = Grpc::Common::prepareHeaders(
+      parent_.cluster_.info()->name(), parent_.service_method_.service()->full_name(),
+      parent_.service_method_.name());
+  headers_message->headers().insertUserAgent().value().setReference(
+      Http::Headers::get().UserAgentValues.EnvoyHealthChecker);
+
+  request_encoder_->encodeHeaders(headers_message->headers(), false);
+
+  if (request_encoder_ == nullptr) {
+    // Stream was synchronously reset inside encodeHeaders() - this can happen if GOAWAY has been
+    // received. Healthcheck error has already been reported by onResetStream() callback.
+    return;
+  }
+
+  grpc::health::v1::HealthCheckRequest request;
+  if (parent_.service_name_.valid())
+    request.set_service(parent_.service_name_.value());
+
+  request_encoder_->encodeData(*Grpc::Common::serializeBody(request), true);
+}
+
+void GrpcHealthCheckerImpl::GrpcActiveHealthCheckSession::onResetStream(Http::StreamResetReason) {
+  bool expect_reset = expect_reset_;
+  resetState();
+
+  if (expect_reset)
+    return;
+
+  ENVOY_CONN_LOG(debug, "connection/stream error health_flags={}", *client_,
+                 HostUtility::healthFlagsToString(*host_));
+
+  // TODO(baranov1ch): according to all HTTP standards, we should check if reason is one of
+  // Http::StreamResetReason::RemoteRefusedStreamReset (which may mean GOAWAY),
+  // Http::StreamResetReason::RemoteReset or Http::StreamResetReason::ConnectionTermination (both
+  // mean connection close), check if connection is not fresh (was used for at least 1 request)
+  // and silently retry request on the fresh connection. This is also true for HTTP/1.1 healthcheck.
+  handleFailure(FailureType::Network);
+
+  if (client_) {
+    // Force connection close on healthcheck errors and timeouts.
+    // Rationale: healthcheck error may mean cluster host is going down gracefully. Technically (
+    // for google-written gRPC server libs), it means that host sends us GOAWAY and PING, waiting
+    // PING ACK to initiate real connection close. But nghttp2 does not respond to pings received
+    // after GOAWAY, and envoy does not handle remote-initiated GOAWAYs yet
+    // (https://github.com/nghttp2/nghttp2/issues/1103).
+    // Altogether, this results in us holding the connection, preventing gRPC server graceful
+    // shutdown until server close timeout expires (typically 1 minute). This is bad. It's better to
+    // drop connection in case of errors than to mess up backend graceful shutdown. Still, all
+    // successful checks will be performed on a single HTTP/2 connection.
```

Comment:
How does any of this differ from other cases where we need to handle GOAWAYs and connection closes on Envoy's data or control planes?

Reply:
Other HTTP/2 cases all go through cluster manager and `Http2::ConnPool` which handles remote goaways implementing `Http::ConnectionCallbacks` by closing the client. I'd better use it here too, thanks for pointing out.

Reply:
It got much cleaner after using `onGoAway` callback